modules/mod-deep-learning/DeepLearningAnalyzer.cpp

```diff
@@ -0,0 +1,232 @@
+
+#include "DeepLearningAnalyzer.h"
+
+#include <cstddef>
+#include <string>
+#include <vector>
+#include <wx/stattext.h>
+
+#include "FileNames.h"
+#include "Shuttle.h"
+#include "ShuttleGui.h"
+#include "WaveTrack.h"
+#include "LabelTrack.h"
+
+#include "effects/LoadEffects.h"
+
+#include <torch/script.h>
+
+// DeepLearningEffectBase implementation
+
+std::string DeepLearningAnalyzer::GetDeepEffectID()
+{ return "waveform-to-labels";}
+
+const ComponentInterfaceSymbol DeepLearningAnalyzer::Symbol
+{ XO("Deep Learning Analyzer") };
+
+// register audio Labeler
+namespace { BuiltinEffectsModule::Registration<DeepLearningAnalyzer> reg; }
+
+DeepLearningAnalyzer::DeepLearningAnalyzer() 
+{ 
+   SetLinearEffectFlag(false); 
+}
+
+DeepLearningAnalyzer::~DeepLearningAnalyzer() 
+{
+}
+
+
+// ComponentInterface implementation
+
+ComponentInterfaceSymbol DeepLearningAnalyzer::GetSymbol() { return Symbol; }
+
+TranslatableString DeepLearningAnalyzer::GetDescription() 
+{
+  return XO("The auto labeler uses deep learning models to "
+         "annotate audio tracks based on their contents automatically."); 
+}
+
+ManualPageID DeepLearningAnalyzer::ManualPage() 
+{
+  return L"Audio_Labeler"; 
+}
+
+// EffectDefinitionInterface implementation
+
+EffectType DeepLearningAnalyzer::GetType() { return EffectTypeAnalyze; }
+
+// Effect implementation
+
+// ProcessOne() takes a track, transforms it to bunch of buffer-blocks,
+// performs a forward pass through the deep model, and writes
+// the output to new tracks.
+bool DeepLearningAnalyzer::ProcessOne(WaveTrack *leader, double tStart, double tEnd) 
+{
+   // get current models labels
+   std::vector<std::string> classList = mModel->GetCard()->labels();
+   mClasses.clear();
+   for (size_t i = 0; i < classList.size(); i++) 
+   {
+      std::cout << classList[i];
+      mClasses.emplace_back(wxString(classList[i]));
+   }
+
+   if (leader == NULL)
+      return false;
+
+   wxString labelTrackName(leader->GetName() + " Labels");
+   std::shared_ptr<AddedAnalysisTrack> labelTrack =
+      AddAnalysisTrack(labelTrackName);
+
+   sampleFormat origFmt = leader->GetSampleFormat();
+   int origRate = leader->GetRate();
+
+   // Initiate processing buffer, most likely shorter than
+   // the length of the selection being processed.
+   Floats buffer{leader->GetMaxBlockSize()};
+
+   // get each of the blocks we will process
+   for (BlockIndex block : GetBlockIndices(leader, tStart, tEnd)) 
+   {
+      // Get a blockSize of samples (smaller than the size of the buffer)
+      sampleCount samplePos = block.first;
+      size_t blockSize = block.second;
+
+      // get a torch tensor from the leader track
+      torch::Tensor input =
+         BuildMultichannelTensor(leader, buffer.get(), samplePos, blockSize);
+
+      // resample!
+      input = mModel->Resample(input, origRate, mModel->GetSampleRate());
+
+      // if we're not doing a multichannel forward pass, downmix
```

Comment:
I observe downmizing after resampling in the Analyzer, but the opposite in the Effect.  Is the difference a mistake?  Or is it significant, and should be commented?  Or is it unimportant?

Reply:
it is unimportant. fixed for consistency. 1756e7ca96f0dcdd5650aaebc950f8390d3dbc3d

Reply:
it is unimportant,  but I fixed it for consistency. 1756e7ca96f0dcdd5650aaebc950f8390d3dbc3d